design-documents/bit-precise-types.rst

```diff
@@ -20,72 +20,327 @@ a different type.
+Given the mention of 512x512 matrices in the comment where we heard of this we
+assume that in general such a library would be CPU-bound code.
+
+Hence we assume that the main consideration here would be performance.
+
+
+Summary of use-case trade-offs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In our estimation, the C to FPGA use case seems to be the most promising.  We
+estimate that the use in this space will be a great majority of the use of this
+new type.
+
+Uses for cryptography, networking, and in order to help the compiler optimize
+certain code seem like they are large enough to consider but not as widespread.
+
+For the C to FPGA use case, the majority of the use is not expected to be seen
+on Arm Architectures.  For helping the compiler optimize code we expect to only
+see bit-precise types with sizes matching that of standard integral types.
+Cryptographic uses are only expected on "large" sizes which are powers of two.
+Networking uses are likely to be using bit-fields for in-memory representations.
+
+All use-cases would have concerns around performance and the familiarity of
+representations.  There does not seem to be a clear choice to prefer one or the
+other.
+
 
 Alignment and sizes
 -------------------
 
+Options and their trade-offs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
 These types must be at least byte-aligned so they are addressable, and at least
-rounded to a byte boundary in size for ``sizeof``.  Since these types have an
-aesthetic similarity to bit-fields one might expect better packing in an array
-of ``_BitInt(24)`` than an array of ``int32_t`` types (i.e. packing as good as a
-byte-array).  However, this would require a low alignment of such types and that
-would mean loading and storing of even "small" sized ``_BitInt``'s crossing
-cache boundaries -- leading to an unnecessary performance hit and hindering any
-atomic operations on these.
+rounded to a byte boundary in size for ``sizeof``.
+
+"Small" regime
+//////////////
+For the "small" regime there are 2 obvious options:
+
+A. Byte alignment.
+B. Alignment and size "as if" stored in the next-largest Fundamental Data Type.
+   (Where the Fundamental Data Types are defined in the relevant PCS documents).
+
+Option ``A`` has the following benefits:
+
+- Better packing in an array of ``_BitInt(24)`` than an array of ``int32_t``.
+  This is more relevant for bit-precise types than others since these types have
+  an aesthetic similarity to bit-fields and hence programmers might expect good
+  packing.
+
+Option ``B`` has the following benefits (both following from the alignment being
+greater than or equal to the size of the object in memory):
+
+- Avoid a performance hit since loading and storing of these "small" sized
+  ``_BitInt``'s will not cross cache boundaries.
+- Atomic loads and stores can be made on these objects.
+- The representation of bit-precise types of the same size as standard integer
+  types will have the same alignment and size in memory.
```

Comment:
That last point is really the most important. The first 2 points are not too convincing since unaligned accesses are practically for free on modern cores, and LSE2 allows atomics on unaligned data.

Reply:
Ooh -- thanks!
I did not realise the cost of unaligned accesses were essentially free (useful to know).

Reply:
Yes, optimization guides mention things like "The Neoverse N2 handles most unaligned accesses without performance penalties.", so you are unlikely to notice a penalty in typical compiled code.

Reply:
It's not just about the atomic RMW operations, you also (I think) want ordinary reads and writes of _BitInt variables to be single-copy-atomic and implemtable with ordinary loads and stores. This implies they should be aligned so as not to cross cache boundaries, and also not packed along with other objects (i.e. no packing a _BitInt(24) together with a _BitInt(8) in a 32-bit word).